### ISSUE TITLE
PyArray_SimpleNewFromData should accept NULL data for 0-sized arrays

### ISSUE THREAD

**Issue by Cwiiis** [numpy/numpy#9923]
If you have an array with a dimension that has length zero, PyArray_SimpleNewFromData will return an error object instead of an empty array. This is inconvenient when using SWIG to bind C++ libraries that may return empty arrays.

Ideally, if a dimension had size zero, the data parameter would accept NULL, as it would be unused.

**Comment by eric-wieser** [comment]
The danger here is that anything computing `data + index` invokes undefined behaviour if `data == NULL`. For instance, `np.array([])[0:]` computes `data + 0`, which in C (not C++) is UB.

**Comment by Cwiiis** [comment]
True, good point - I think this would be nice, but understandable if it isn't feasible - if not, it might be worth mentioning in the documentation(?)

**Comment by eric-wieser** [comment]
I suppose a super easy fix would be to set `data = malloc(1)` if `data == NULL` and `PyArray_SIZE(self) == 0`.

Also, I think we already hit UB for `np.empty((10,0))[5:]`...
